test/StreamJsonRpc.Tests/MarshalableProxyTests.cs

```diff
@@ -0,0 +1,522 @@
+            this.TimeoutToken));
+        return new WeakReference(marshalable);
+    }
+
+    public class Server : IServer
+    {
+        internal AsyncManualResetEvent ReturnedMarshalableDisposed { get; } = new AsyncManualResetEvent();
+
+        internal Data? ReturnedMarshalable { get; set; }
+
+        internal IMarshalable? ReceivedProxy { get; set; }
+
+        public Task<IMarshalable?> GetMarshalableAsync(bool returnNull) => Task.FromResult<IMarshalable?>(returnNull ? null : this.ReturnedMarshalable = new Data(() => this.ReturnedMarshalableDisposed.Set()));
+
+        public Task<IGenericMarshalable<int>?> GetGenericMarshalableAsync(bool returnNull) => Task.FromResult<IGenericMarshalable<int>?>(returnNull ? null : this.ReturnedMarshalable = new Data(() => this.ReturnedMarshalableDisposed.Set()));
+
+        public Task AcceptProxyAsync(IMarshalable marshalable, bool dispose = true)
+        {
+            this.ReceivedProxy = marshalable;
+            if (dispose)
+            {
+                marshalable.Dispose();
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task AcceptGenericProxyAsync(IGenericMarshalable<int> marshalable, bool dispose = true)
+        {
+            this.ReceivedProxy = marshalable;
+            if (dispose)
+            {
+                marshalable.Dispose();
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task AcceptProxyContainerAsync(ProxyContainer<IMarshalable> marshalableContainer, bool dispose = true)
+        {
+            this.ReceivedProxy = marshalableContainer.Marshalable;
+            if (dispose)
+            {
+                marshalableContainer.Marshalable?.Dispose();
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task AcceptGenericProxyContainerAsync(ProxyContainer<IGenericMarshalable<int>> marshalableContainer, bool dispose = true)
+        {
+            this.ReceivedProxy = marshalableContainer.Marshalable;
+            if (dispose)
+            {
+                marshalableContainer.Marshalable?.Dispose();
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task<int> AcceptDataAsync(Data data) => Task.FromResult(data.Value);
+
+        public Task<Data> ReturnDataAsync(int value) => Task.FromResult(new Data { Value = value });
+
+        public Task<int> AcceptDataContainerAsync(DataContainer dataContainer) => Task.FromResult(dataContainer.Data?.Value ?? 0);
+
+        public Task AcceptNonDisposableMarshalableAsync(INonDisposableMarshalable nonDisposable) => Task.CompletedTask;
+
+        public Task AcceptMarshalableWithPropertiesAsync(IMarshalableWithProperties marshalableWithProperties) => Task.CompletedTask;
+
+        public Task AcceptMarshalableWithEventsAsync(IMarshalableWithEvents marshalableWithEvents) => Task.CompletedTask;
+    }
+
+    [DataContract]
+    public class ProxyContainer<T>
```

Comment:
How about a class which implements a marshalable interface and includes `[DataContract]` but is passed by its concrete type rather than as the interface? In such a case, it should be passed by-value instead of by-reference.

Reply:
`IMarshalableDataAsArg_ShouldSerialize` and `IMarshalableDataAsObjectWithinArg_ShouldSerialize` should already test that. We were already testing this case for `IDisposable` and I duplicated all the `IDisposable` tests here.

Reply:
Ah yes, I see. Thanks.